### PR TITLE
Add credential_revoked state

### PR DESCRIPTION
**Pull request by DaevMithran**
Pull request for [Issue: 1539](https://github.com/hyperledger/aries-cloudagent-python/issues/1539)
@swcurran  Added the credential_revoked state. 
For now I have considered only issue-credential 1.0, I should check for the record in V20CredentialExchange too.
Will do it soon. If you have any suggestions on my approach please comment below.

Questions:
- Is it required raise an error if the credential record doesn't exist? 
- The credential state should be changed only after publishing revocation.
   When publish is set to pending, what is the best way to implement this?. I'm planning to create retrieve credential records using rev_reg_id and cred_rev_id

### CODE REVIEW

**Comment by ianco** [comment]
Typically the credential exchange records will get deleted after the credential is issued, so in most use cases there will be no credential exchange record when the credential is revoked.

I'm not sure if it makes sense to add a status or tag on the credential itself with the status (issued vs revoked)?

**Comment by swcurran** [comment]
> Typically the credential exchange records will get deleted after the credential is issued, so in most use cases there will be no credential exchange record when the credential is revoked.
> 
> I'm not sure if it makes sense to add a status or tag on the credential itself with the status (issued vs revoked)?

While "typically" they will get deleted, not always, and the intention of this PR is to updated the non-deleted protocol state objects if they are not deleted.

**Comment by ianco** [comment]
> * Is it required raise an error if the credential record doesn't exist?

It shouldn't raise an error.  The credential exchange record can be deleted, and could even be deleted in between the revocation and publishing events.

> * The credential state should be changed only after publishing revocation.
>   When publish is set to pending, what is the best way to implement this?. I'm planning to create retrieve credential records using rev_reg_id and cred_rev_id

Not completely sure of the intent, but you may want to change the status to `revocation_pending` if publishing the revocation status will be delayed, otherwise it looks like the credential is still in `issued` status.  Then, when you publish revocations, you can just flip the status from `revocation_pending` to `revoked`.

Also note that there are some tests failing on this PR (formatting and unit tests).

**Comment by DaevMithran** [comment]
Hi @swcurran  @ianco 
For the current approach I added a revoke_credentials method in Credential Manager which retrieves credential records using revoc_reg_id and revocation_id and changes their state to credential_revoked.

This approach supports credentials records issued after this merge. As the previous credential records did not have tags to be retrieved using revoc_reg_id and revocation_id

File: aries_cloudagent/protocols/issue_credential/v1_0/manager.py [hunk 1]
```diff
@@ -893,6 +895,33 @@ async def receive_credential_ack(
 
         return cred_ex_record
 
+    async def revoke_credentials(
```
Comment (ianco) [inline aries_cloudagent/protocols/issue_credential/v1_0/manager.py:898]: I suggest re-naming this method to something like "set_credential_revocation_state" to be more clear on what it's doing (i.e. it's not revoking the credentials, just updating the state).

**Review by ianco — CHANGES_REQUESTED** [review]
One (fairly minor) comment, plus this functionality should also be provided for the V20 cred exchange.

File: aries_cloudagent/protocols/issue_credential/v1_0/manager.py (hunk 1 above)
Comment (DaevMithran) [reply aries_cloudagent/protocols/issue_credential/v1_0/manager.py]: Agree. Changed this

**Review by ianco — APPROVED** [review]

**Comment by shaangill025** [comment]
> * This approach supports credentials records issued after this merge. As the previous credential records did not have tags to be retrieved using revoc_reg_id and revocation_id

I don't see any changes to `TAG_NAMES` [here](https://github.com/hyperledger/aries-cloudagent-python/blob/f94184dfc85729b0e0ecce5aa84910feaab36986/aries_cloudagent/protocols/issue_credential/v1_0/models/credential_exchange.py#L37) and [here](https://github.com/hyperledger/aries-cloudagent-python/blob/f94184dfc85729b0e0ecce5aa84910feaab36986/aries_cloudagent/protocols/issue_credential/v2_0/models/cred_ex_record.py#L37)

**Comment by DaevMithran** [comment]
> > * This approach supports credentials records issued after this merge. As the previous credential records did not have tags to be retrieved using revoc_reg_id and revocation_id
> 
> I don't see any changes to `TAG_NAMES` [here](https://github.com/hyperledger/aries-cloudagent-python/blob/f94184dfc85729b0e0ecce5aa84910feaab36986/aries_cloudagent/protocols/issue_credential/v1_0/models/credential_exchange.py#L37) and [here](https://github.com/hyperledger/aries-cloudagent-python/blob/f94184dfc85729b0e0ecce5aa84910feaab36986/aries_cloudagent/protocols/issue_credential/v2_0/models/cred_ex_record.py#L37).  I am just confused by the above comment regarding `cred_ex_record` tags. Just to confirm in regards to `cred_ex_record` backward compatibility, there should be no breaking change, and only that `cred_ex_record` state will be updated as `credential_revoked` upon revocation after the merge?

Hi @shaangill025 initially I tried the tags approach but V20Cred_ex_record doesn't include values of rev_reg_id and cred_rev_id so went with a different approach. 

- First I retrieve IssuerCredRevRecord using rev_reg_id and cred_rev_id
- extract the cred_ex_id from the record 
- retrieve credential record using cred_ex_id
- update state of credential record

**Pull request merged**